browser/brave_cosmetic_resources_tab_helper.cc

```diff
@@ -0,0 +1,499 @@
+      size_t last_dot_pos = host.find(".", found_pos + 1);
+      if (last_dot_pos == std::string::npos) {
+        return false;
+      }
+      if (host.find(".", last_dot_pos + 1) == std::string::npos) {
+        return true;
+      }
+    }
+  }
+
+  return false;
+}
+
+}  // namespace
+
+
+BraveCosmeticResourcesTabHelper::BraveCosmeticResourcesTabHelper(
+    content::WebContents* contents)
+    : WebContentsObserver(contents),
+    enabled_1st_party_cf_filtering_(false) {
+  if (BraveCosmeticResourcesTabHelper::observing_script_->empty()) {
+    *BraveCosmeticResourcesTabHelper::observing_script_ =
+        LoadDataResource(kCosmeticFiltersGenerated[0].value);
+  }
+}
+
+BraveCosmeticResourcesTabHelper::~BraveCosmeticResourcesTabHelper() {
+}
+
+void BraveCosmeticResourcesTabHelper::GetUrlCosmeticResourcesOnUI(
+    content::GlobalFrameRoutingId frame_id, const std::string& url,
+    bool do_non_scriplets, std::unique_ptr<base::ListValue> resources) {
+  if (!resources) {
+    return;
+  }
+  for (auto i = resources->GetList().begin();
+      i < resources->GetList().end(); i++) {
+    base::DictionaryValue* resources_dict;
+    if (!i->GetAsDictionary(&resources_dict)) {
+      continue;
+    }
+    std::string to_inject;
+    resources_dict->GetString("injected_script", &to_inject);
+    auto* frame_host = content::RenderFrameHost::FromID(frame_id);
+    if (!frame_host)
+      return;
+    if (do_non_scriplets) {
+      Profile* profile = Profile::FromBrowserContext(
+          web_contents()->GetBrowserContext());
+      enabled_1st_party_cf_filtering_ =
+          brave_shields::IsFirstPartyCosmeticFilteringEnabled(
+              HostContentSettingsMapFactory::GetForProfile(profile),
+                  GURL(url));
+      bool generichide = false;
+      resources_dict->GetBoolean("generichide", &generichide);
+      std::string pre_init_script =
+        "(function() {"
+        "if (window.hide1pContent === undefined) {"
+          "window.hide1pContent = ";
+      if (enabled_1st_party_cf_filtering_) {
+        pre_init_script += "true";
+      } else {
+        pre_init_script += "false";
+      }
+      pre_init_script += ";"
+        "}"
+        "if (window.generichide === undefined) {"
+          "window.generichide = ";
+      if (generichide) {
+        pre_init_script += "true";
+      } else {
+        pre_init_script += "false";
+      }
+      pre_init_script += ";"
+        "}"
+        "})();";
+      frame_host->ExecuteJavaScriptInIsolatedWorld(
+        base::UTF8ToUTF16(pre_init_script),
+        base::NullCallback(), ISOLATED_WORLD_ID_CHROME_INTERNAL);
+      frame_host->ExecuteJavaScriptInIsolatedWorld(
+        base::UTF8ToUTF16(*BraveCosmeticResourcesTabHelper::observing_script_),
+        base::NullCallback(), ISOLATED_WORLD_ID_CHROME_INTERNAL);
+    }
+    if (to_inject.length() > 1) {
+      std::string script =
+        "(function() {"
+          "let script;"
+          "let text = " + to_inject + ";"
```

Comment:
if `to_inject` is from a third party maintainer, i recommend adding some validation here to avoid XSS (for instance to_inject = `null; doEvilThing();`). 

one option is to just run the chromium C++ equivalent of JSON.parse on `to_inject` and throw an error if it doesn't parse correctly. i think this ensures that `to_inject` will correspond to an Object, Array, string, number, boolean, or null value in js.

Reply:
I'm going to check on something similar like we do with rules here